handlers/handlers.go

```diff
@@ -28,86 +27,165 @@ type RenderClient interface {
+		log.Event(ctx, "getting search response from client failed", log.Error(err), log.ERROR)
+		setStatusCode(w, req, err)
+		return
+	}
+
+	// TO-DO: Until API handles aggregration on datatypes (e.g. bulletins, article), we need to make a second request
+
+	err = validateCurrentPage(ctx, validatedQueryParams, resp.Count)
+	if err != nil {
+		log.Event(ctx, "unable to validate current page", log.Error(err), log.ERROR)
+		setStatusCode(w, req, err)
+		return
+	}
+
+	categories, err := getCategoriesTypesCount(ctx, apiQuery, searchC)
+	if err != nil {
+		log.Event(ctx, "getting categories, types and its counts failed", log.Error(err), log.ERROR)
+		setStatusCode(w, req, err)
+		return
+	}
+
+	err = getSearchPage(w, req, cfg, rendC, validatedQueryParams, categories, resp)
+	if err != nil {
+		log.Event(ctx, "getting search page failed", log.Error(err), log.ERROR)
+		setStatusCode(w, req, err)
+		return
 	}
-	log.Event(req.Context(), "setting-response-status", log.Error(err), log.ERROR)
-	w.WriteHeader(status)
 }
 
-var marshal = func(v interface{}) ([]byte, error) {
-	return json.Marshal(v)
+// validateCurrentPage checks if the current page exceeds the total pages which is a bad request
+func validateCurrentPage(ctx context.Context, validatedQueryParams data.SearchURLParams, resultsCount int) (err error) {
+
+	if resultsCount > 0 {
+		totalPages := data.GetTotalPages(validatedQueryParams.Limit, resultsCount)
+
+		if validatedQueryParams.CurrentPage > totalPages {
+			err = errs.ErrPageExceedsTotalPages
+			log.Event(ctx, "current page exceeds total pages", log.Error(err), log.ERROR)
+
+			return err
+		}
+	}
+
+	return nil
 }
 
-var writeResponse = func(w http.ResponseWriter, templateHTML []byte) (int, error) {
-	return w.Write(templateHTML)
+// getCategoriesTypesCount removes the filters and communicates with the search api again to retrieve the number of search results for each filter categories and subtypes
+func getCategoriesTypesCount(ctx context.Context, apiQuery url.Values, searchC SearchClient) ([]data.Category, error) {
+	//Remove filter to get count of all types for the query from the client
+	apiQuery.Del("content_type")
+
+	countResp, err := searchC.GetSearch(ctx, apiQuery)
+	if err != nil {
+		log.Event(ctx, "getting search query count from client failed", log.Error(err), log.ERROR)
+		return nil, err
+	}
+
+	categories := data.GetCategories()
+
+	setCountToCategories(ctx, countResp, categories)
+
+	return categories, nil
+}
+
+func setCountToCategories(ctx context.Context, countResp searchC.Response, categories []data.Category) {
+	for _, responseType := range countResp.ContentTypes {
+		foundFilter := false
+
+	categoryLoop:
+		for i, category := range categories {
+			for j, contentType := range category.ContentTypes {
+				for _, subType := range contentType.SubTypes {
+					if responseType.Type == subType {
+						categories[i].Count += responseType.Count
+						categories[i].ContentTypes[j].Count += responseType.Count
```

Comment:
You need to be careful when accessing array element by index.  I think this specific case is OK because `categories` is an array of **non pointer** types so each element of the array cannot be `nil`. No action to take here but I thought it would highlight the potential risk so you are aware of it.

If categories was a list of pointer types i.e. `[]*data.Category` you would need to be careful and not assume an entry is not nil.

For example if you had in the input
```
[]*data.Category{nil, nil, nil}
```
The `for` loop would iterate over the array entries but it would nil pointer error when it tries to do `categories[i].Count`.

This is essentially what happened in last weeks issue and what we discussed on the postmortem call. The simple solution to this is to _alway assume it could be nil_ and do a nil check before you access it.